File: C++/代码随想录/Dynamic_Programming/Subsequence_problem/53. Maximum Subarray.py
def maxSubArray(nums):
    n = len(nums)
    if n == 0:
        return 0

    max_sum = current_sum = nums[0]
    for i in range(1, n):
        current_sum = max(nums[i], current_sum + nums[i])
        max_sum = max(max_sum, current_sum)

    return max_sum

# 测试用例
print(maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))  # 输出应该是 6

File: C++/代码随想录/Dynamic_Programming/Subsequence_problem/53. Maximum Subarray.cpp
// 53. Maximum Subarray
// https://leetcode.com/problems/maximum-subarray/

#include <iostream>
#include <vector>
#include <algorithm>

int maxSubArray(std::vector<int>& nums) {
    int n = nums.size();
    std::vector<int> dp(n, 0);
    dp[0] = nums[0];
    int maxSum = dp[0];

    for (int i = 1; i < n; i++) {
        dp[i] = std::max(nums[i], nums[i] + dp[i - 1]);
        maxSum = std::max(maxSum, dp[i]);
    }

    return maxSum;
}

int main() {
    std::vector<std::vector<int>> testCases = {
        {-2, 1, -3, 4, -1, 2, 1, -5, 4},
        {1},
        {5, 4, -1, 7, 8}
    };

    for (int i = 0; i < testCases.size(); i++) {
        std::vector<int> nums = testCases[i];

        int maxSubArrSum = maxSubArray(nums);

        std::cout << "Test case " << (i + 1) << ": ";
        std::cout << "The maximum sum of the subarray is: " << maxSubArrSum << std::endl;
    }

    return 0;
}